cql2pgjson/src/main/java/org/folio/cql2pgjson/util/DbSchemaUtils.java

```diff
@@ -54,4 +65,99 @@ public static Index getIndex(String cqlIndex, List<Index> indexes) {
     return null;
   }
 
+  /**
+   * Get {@link Table} for given table name in {@link Schema}
+   *
+   * @param dbSchema
+   * @param tableName
+   * @return table; null if not found
+   */
+  public static Table getTable(Schema dbSchema, String tableName) {
+    for (Table table : dbSchema.getTables()) {
+      if (tableName.equals(table.getTableName())) {
+        return table;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Find a list of {@link ForeignKeys} from source table to target table alias
+   *
+   * @param dbSchema
+   * @param srcTabName
+   * @param targetTabAlias
+   *
+   * @return a list of {@link ForeignKeys}; empty if none found
+   */
+  public static List<DbFkInfo> findForeignKeysFromSourceTableToTargetAlias(Schema dbSchema, String srcTabName,
+      String targetTabAlias) {
+    return findForeignKeys(dbSchema, srcTabName, targetTabAlias, true);
+  }
+
+  private static List<DbFkInfo> findForeignKeys(Schema dbSchema, String srcTabName, String targetTabName,
+      boolean useTargetAlias) {
+    List<DbFkInfo> list = new ArrayList<>();
+    Table srcTab = getTable(dbSchema, srcTabName);
+    if (srcTab == null || srcTab.getForeignKeys() == null) {
+      return list;
+    }
+    // direct FK
+    for (ForeignKeys fk : srcTab.getForeignKeys()) {
+      String targetName = useTargetAlias ? fk.getTargetTableAlias() : fk.getTargetTable();
+      if (targetTabName.equals(targetName)) {
+        list.add(new DbFkInfo(srcTab.getTableName(), fk.getFieldName(), fk.getTargetTable()));
+        return list;
+      }
+    }
+    // find the shortest path
+    for (ForeignKeys fk : srcTab.getForeignKeys()) {
+      updateFkList(list, dbSchema, srcTab, fk, targetTabName, useTargetAlias);
+    }
+    return list;
+  }
+
+  private static void updateFkList(List<DbFkInfo> list, Schema dbSchema, Table srcTab, ForeignKeys fk,
+      String targetTabName, boolean useTargetAlias) {
+    List<DbFkInfo> childList = findForeignKeys(dbSchema, fk.getTargetTable(), targetTabName, useTargetAlias);
+    if (!childList.isEmpty()) {
+      if (!list.isEmpty() && (list.size() > (childList.size() + 1))) {
+        list.clear();
+      }
+      list.add(new DbFkInfo(srcTab.getTableName(), fk.getFieldName(), fk.getTargetTable()));
+      list.addAll(childList);
+    }
+  }
+
+  /**
+   * Find a list of {@link ForeignKeys} from source table alias and target table.
+   *
+   * @param dbSchema
+   * @param srcTabAlias
+   * @param targetTabName
+   * @return list; empty if none found
+   */
+  public static List<DbFkInfo> findForeignKeysFromSourceAliasToTargetTable(Schema dbSchema, String srcTabAlias,
+      String targetTabName) {
+    List<DbFkInfo> list = new ArrayList<>();
+    for (Table table : dbSchema.getTables()) {
+      if (table.getForeignKeys() == null) {
+        continue;
+      }
+      for (ForeignKeys fk : table.getForeignKeys()) {
+        String tabAlias = fk.getTableAlias() == null ? table.getTableName() : fk.getTableAlias();
```

Comment:
This is contrary to what we decided in RMB-395.
We decided that if tableAlias is null then there will be no sub query at all and that we don't want tableName as a fall back because that will lead to unwanted results:
a) The table name holdings_records is not suitable, we want camel case.
b) There may be a name clash in that there exists a field that is that same as the table name.

Reply:
I don't see that being decided at all. Primarily an optional feature to avoid name clashes.